### PR TITLE
Fix handling of errors.data field transformation

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

See title - also includes updating documentation on how to implement v2 log.go library

### How to review

Check changes make sense

### Who can review

Anyone

### CODE REVIEW

**Review by redhug1 — COMMENTED** [review]
error.go, line 48, typo: strack

middleware_test.go:
'req' might have 'nil' or other unexpected value as its corresponding error variable might be not 'nil' (err not checked after line 147)
Variable 'http' collides with imported package name, line 163
Variable 'http' collides with imported package name, line 189
Redundant type conversion, line 114
Typo: In word 'respones', line 78

log_test.go:
Unhandled error, lines 159,163,166
Redundant alias, line 11
Redundant parentheses, 372

log.go:
Function 'print' collides with the builtin function, line 270
Method call 'err.Error()' might lead to a nil pointer dereference, line 291
Redundant alias, line 14 and 15
Typo: In word 'marshallable', line 240

http_test.go :
Variable 'http' collides with imported package name, line 15, 24, 49, 57, 65

http.go :
Exported function with the unexported return type, line 46

event_mock.go 👍 
Unused parameter 'severity', line 16

error.go, line 51, same as above

data.go, line 8, Typo: In word 'indexable'

auth.go :
Exported function with the unexported return type, line 23

Code coverage:
Is it worth creating a test to cover lines 68 and 69 of error.go - for the 'nil' case ?

File: log/error_test.go [hunk 1]
```diff
@@ -38,7 +30,7 @@ func TestFormatErrorsFunc(t *testing.T) {
 		So(origin.File, ShouldEndWith, "log.go/log/error_test.go")
 
 		// If this test fails, check the `errEventData := Error(err).(*EventErrors)` line is still line 34!
-		So(origin.Line, ShouldEqual, 34)
+		So(origin.Line, ShouldEqual, 26)
```
Comment (redhug1) [inline log/error_test.go:33]: Whats with the magic number of 26
Shouldn't there be const strings and taking lengths of those ?

Comment (nshumoogum) [reply log/error_test.go:33]: This is the line number of when an event is thrown, as we update this file the test might move line number and this is reflected in this value changing - hence a previous commit of mine was to move this test to the top of the test (less likely to change).

Comment (nshumoogum) [reply log/error_test.go:33]: I do need to update the comment

**Comment by nshumoogum** [comment]
@redhug1 - I've addressed most of your suggestions in latest commits

re. test scenario for checking if error is nil - I've refactored the code slightly as the test discovered another issue and would panic with error `invalid memory address or nil pointer dereference`

The unexported return types on exported functions is valid - we do not want the caller to be able to edit or create their own option - in go this use is discouraged but does not mean it should not be done when there is a valid reason

**Comment by redhug1** [comment]
Hi Nathan,
You say:

“The unexported return types on exported functions is valid - we do not want the caller to be able to edit or create their own option - in go this use is discouraged but does not mean it should not be done when there is a valid reason“
When there are exceptions, it helps to have a comment explaining why its been done the way it has to save others ‘spotting’ what i’ve spotted ...

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

**Comment by nshumoogum** [comment]
@redhug1 - there is already a comment explaining this above the definition of this new option type in log.go

**Review by redhug1 — APPROVED** [review]
LGTM

**Pull request merged**